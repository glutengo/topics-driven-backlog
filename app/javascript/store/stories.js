import Vue from 'vue';

export default {
    namespaced: true,
    state: {
        initialised: false,
        statusMap: [
            {name: 'open', css: 'badge-dark'},
            {name: 'in progress', css: 'badge-warning'},
            {name: 'closed', css: 'badge-success'},
            {name: 'cancelled', css: 'badge-danger'}
        ],
        data: []
    },
    mutations: {
        /**
         * Set one or all stories
         * @param state
         * @param payload
         */
        set: (state, payload) => {
            if (payload.stories) {
                Vue.set(state, 'data', payload.stories);
            }

            if (payload.story) {
                const isPresent = state.data
                    .find(story => story.id === payload.story.id);

                if (isPresent) {
                    Vue.set(state, 'data', state.data.map(story =>
                        (story.id === payload.story.id ? payload.story : story),
                    ));
                } else {
                    state.data.push(payload.story);
                }
            }
        },
    },
    actions: {
        /**
         * Initialises the stories store
         * @param state
         * @param dispatch
         */
        init({state, dispatch}) {
            if (!state.initialised) {
                dispatch('fetch').then(() => {
                    state.initialised = true;
                });
            }
        },

        /**
         * Fetches all stories from the API and updates store
         * @param commit
         * @returns {Promise}
         */
        fetch({commit}) {
            return new Promise((resolve, reject) => {
                Vue.http.get('/stories').then((response) => {
                    commit('set', {
                        stories: response.body,
                    });

                    resolve(response.body);
                }, reject);
            });
        },

        /**
         * Saves a story to the API and updates store
         * @param commit
         * @param payload (story)
         * @returns {Promise}
         */
        save({commit}, payload) {
            return new Promise((resolve, reject) => {
                Vue.http.post('/stories', payload.story).then((response) => {
                    commit('set', {
                        story: response.body,
                    });

                    resolve(response.body);
                }, reject);
            });
        },

        /**
         * Patches a given field of a given story with a given value
         * @param commit
         * @param payload (id, field, value)
         * @returns {Promise}
         */
        patch({commit}, payload) {
            return new Promise((resolve, reject) => {
                const data = {};
                data[payload.field] = payload.value;

                Vue.http.patch(`/stories/${payload.id}`, data).then((response) => {
                    commit('set', {
                        story: response.body,
                    });

                    resolve(response.body);
                }, reject);
            });
        },

        /**
         * Saves order information to a story and triggers a new fetch of stories
         * @param getters
         * @param dispatch
         * @param payload
         * @returns {Promise}
         */
        reorder({getters, dispatch}, payload) {
            const story = getters.all[payload.oldIndex];

            return new Promise((resolve, reject) => {
                if (!story) {
                    return reject();
                }

                Vue.http.patch(`/stories/${story.id}`, {position: payload.newIndex}).then((response) => {
                    dispatch('fetch').then(resolve, reject);
                }, reject);
            });
        }
    },
    getters: {
        /**
         * All stories
         * @param state
         */
        all: state => state.initialised
            ? state.data.sort((a, b) => a.position > b.position)
            : [],

        /**
         * Find a story by identifier
         * @param state
         */
        byIdentifier: state => value => state.initialised
            ? state.data.find(s => s.identifier === value)
            : null,
    }
}