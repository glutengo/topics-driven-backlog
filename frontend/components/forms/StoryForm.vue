<template>
    <BForm 
        @submit.prevent="submit" 
        class="story-form">

        <BRow>
            <BCol md="10">
                <BFormGroup 
                    label="Story *" 
                    label-for="story-title">

                    <BFormInput 
                        id="story-title" 
                        v-model="data.title" 
                        required/>
                </BFormGroup>
            </BCol>
            <BCol md="2">
                <BFormGroup 
                    label="Story points" 
                    label-for="story-points">

                    <BFormInput 
                        id="story-points" 
                        type="number" 
                        min="0" 
                        v-model="data.points"/>
                </BFormGroup>
            </BCol>
        </BRow>

        <BFormGroup
                label="Topic"
                label-for="story-topic">

            <BFormSelect
                    id="story-topic"
                    v-model="data.topic_id">

                <option :value="null" selected>No topic</option>

                <option
                    v-for="topic in topics"
                    :value="topic.id"
                    :key="topic.id">

                    {{ topic.title }}
                </option>
            </BFormSelect>
        </BFormGroup>

        <BFormGroup 
            label="Notes" 
            label-for="story-description">

            <BFormTextarea 
                id="story-description" 
                v-model="data.description"/>
        </BFormGroup>

        <BButton 
            type="submit" 
            variant="primary"
            :class="{ 'is-disabled': isInvalid}">Save</BButton>

        <BButton 
            type="button" 
            variant="grey" 
            @click="cancel">Cancel</BButton>
    </BForm>
</template>

<script>
import { required, numeric, minValue } from 'vuelidate/lib/validators';
import BForm from '@bootstrap/form/form';
import BFormGroup from '@bootstrap/form-group/form-group';
import BFormInput from '@bootstrap/form-input/form-input';
import BFormTextarea from '@bootstrap/form-textarea/form-textarea';
import BFormSelect from '@bootstrap/form-select/form-select';
import BRow from '@bootstrap/layout/row';
import BCol from '@bootstrap/layout/col';
import BButton from '@bootstrap/button/button';
import BaseForm from '@/components/forms/BaseForm';

export default {
    name: 'StoryForm',
    components: { BForm, BFormGroup, BFormInput, BFormTextarea, BFormSelect, BRow, BCol, BButton },
    extends: BaseForm,
    props: {
        project: {
            type: Object,
            default: null,
        },
    },
    computed: {
        topics() {
            return this.project
                ? this.$store.getters['topics/all'](this.project.course_id)
                : [];
        },
    },
    validations: {
        data: {
            title: { required },
            points: { numeric, minValue: minValue(0) },
        },
    },
};
</script>