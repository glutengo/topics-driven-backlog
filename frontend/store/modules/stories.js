import { generate } from '@/store/base/resource-module/index';
import resources from '@/store/resources';

const resourceModule = generate({
    resource: resources.STORIES,
    parent: resources.PROJECTS,
    state: {
        STATUS: { OPEN: 0, IN_PROGRESS: 1, CLOSED: 2, CANCELLED: 3 },
        statusMap: [
            { name: 'open', css: 'badge-dark' },
            { name: 'in progress', css: 'badge-warning' },
            { name: 'closed', css: 'badge-success' },
            { name: 'cancelled', css: 'badge-danger' },
        ],
    },
    template: () => {
        return {
            topic_id: null,
        };
    },
});

export default resourceModule;
