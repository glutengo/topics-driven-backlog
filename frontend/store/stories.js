import Vue from 'vue';

export default {
    namespaced: true,
    state: {
        initialised: false,
        STATUS: { OPEN: 0, IN_PROGRESS: 1, CLOSED: 2, CANCELLED: 3 },
        statusMap: [
            { name: 'open', css: 'badge-dark' },
            { name: 'in progress', css: 'badge-warning' },
            { name: 'closed', css: 'badge-success' },
            { name: 'cancelled', css: 'badge-danger' },
        ],
        data: [],
    },
    mutations: {
        /**
         * Set one or all stories
         * @param {object} state
         * @param {object} payload
         */
        set: (state, payload) => {
            if (payload.stories) {
                Vue.set(state, 'data', payload.stories);
            }

            if (payload.story) {
                const isPresent = state.data.find(
                    story => story.id === payload.story.id
                );

                if (isPresent) {
                    Vue.set(
                        state,
                        'data',
                        state.data.map(
                            story =>
                                story.id === payload.story.id
                                    ? payload.story
                                    : story
                        )
                    );
                } else {
                    state.data.push(payload.story);
                }
            }
        },

        /**
         * Remove story from data set
         * @param {object} state
         * @param {object} payload
         */
        remove: (state, payload) => {
            Vue.set(
                state,
                'data',
                state.data.filter(story => story.id !== payload.id)
            );
        },
    },
    actions: {
        /**
         * Initialises the stories store
         * @param {object} state
         * @param {function} dispatch
         */
        init({ state, dispatch }) {
            if (!state.initialised) {
                dispatch('fetch').then(() => {
                    state.initialised = true;
                });
            }
        },

        /**
         * Fetches all stories from the API and updates store
         * @param {function} commit
         * @returns {Promise}
         */
        fetch({ commit }) {
            return new Promise((resolve, reject) => {
                Vue.http.get('/stories').then(response => {
                    commit('set', {
                        stories: response.body,
                    });

                    resolve(response.body);
                }, reject);
            });
        },

        /**
         * Saves a story to the API and updates store
         * @param {function} commit
         * @param {object} payload (story)
         * @returns {Promise}
         */
        save({ commit }, payload) {
            return new Promise((resolve, reject) => {
                Vue.http.post('/stories', payload.story).then(response => {
                    commit('set', {
                        story: response.body,
                    });

                    resolve(response.body);
                }, reject);
            });
        },

        /**
         * Patches a given field of a given story with a given value or all given values
         * @param {function} commit
         * @param {object} payload (id, field, value, values)
         * @returns {Promise}
         */
        patch({ commit, dispatch }, payload) {
            return new Promise((resolve, reject) => {
                let values = {};

                if (payload.values) {
                    values = payload.values;
                } else {
                    values[payload.field] = payload.value;
                }

                Vue.http
                    .patch(`/stories/${payload.id}`, values)
                    .then(response => {
                        if (payload.fetch) {
                            dispatch('fetch').then(resolve, reject);
                        } else {
                            commit('set', {
                                story: response.body,
                            });

                            resolve(response.body);
                        }
                    }, reject);
            });
        },

        /**
         * Deletes a given story
         * @param {function} commit
         * @param {object} payload (id)
         * @returns {Promise}
         */
        delete({ commit }, payload) {
            return new Promise((resolve, reject) => {
                Vue.http.delete(`/stories/${payload.id}`).then(() => {
                    commit('remove', payload);

                    resolve();
                }, reject);
            });
        },
    },
    getters: {
        /**
         * All stories
         * @param {object} state
         * @returns {array}
         */
        all: state => (state.initialised ? state.data : []),

        /**
         * Find a story by identifier
         * @param {object} state
         * @returns {(object|null)}
         */
        byIdentifier: state => value =>
            state.initialised
                ? state.data.find(s => s.identifier === value)
                : null,

        /**
         * Find stories by field and value
         * @param {object} state
         * @returns {array}
         */
        find: state => (field, value) =>
            state.initialised ? state.data.filter(s => s[field] === value) : [],
    },
};
