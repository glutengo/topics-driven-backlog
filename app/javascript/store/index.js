import Vue from 'vue';
import Vuex from 'vuex';

import stories from './stories';

Vue.use(Vuex);

export default new Vuex.Store({
    modules: { stories },
    actions: {
        init({ dispatch }) {
            dispatch('stories/init');
        },
    }
});