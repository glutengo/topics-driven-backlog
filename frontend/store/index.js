import Vue from 'vue';
import Vuex from 'vuex';
import moment from 'moment';

import courses from '@/store/modules/courses';
import projects from '@/store/modules/projects';
import sprints from '@/store/modules/sprints';
import stories from '@/store/modules/stories';

Vue.use(Vuex);

export const state = {
    initialised: false,
    currentDate: moment().format('YYYY-MM-DD'),
};

export const actions = {
    /**
     * Initialise store and submodules
     * @param {object} state
     * @param {function} dispatch
     */
    init({ state, dispatch }) {
        dispatch('courses/init');
        state.initialised = true;
    },
};

export default new Vuex.Store({
    modules: { courses, projects, sprints, stories },
    state,
    actions,
});
