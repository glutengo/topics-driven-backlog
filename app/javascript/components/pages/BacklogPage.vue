<template>
    <section id="backlog-page">
        <h1>Backlog</h1>

        <table class="table table-striped">
            <thead>
                <tr>
                    <th>&nbsp;</th>
                    <th>Identifier</th>
                    <th>Story</th>
                    <th>Story&nbsp;points</th>
                    <th>Operations</th>
                </tr>
            </thead>
            <tbody v-sortable="{handle: '.js-drag-drop', onEnd: saveOrder}" >
                <tr is="story" v-for="story in stories" :key="story.id" :data="story" class="mb-3" view="backlog"></tr>
            </tbody>
        </table>

        <story-form v-if="showForm" v-model="newStory" @cancel="showForm = false" @submit="save"></story-form>
        <b-button v-else type="button" variant="primary" @click="showForm = true">Add story</b-button>
    </section>
</template>

<script>
    import {required, numeric, minValue} from 'vuelidate/lib/validators';
    import Story from "../elements/Story.vue";
    import StoryForm from "../forms/StoryForm.vue";

    export default {
        components: {StoryForm, Story},
        data() {
            return {
                showForm: false,
                newStory: {},
            };
        },
        computed: {
            stories() {
                return this.$store.getters['stories/all']
                    .filter(s => s.status === this.$store.state.stories.STATUS.OPEN)
                    .sort((a, b) => a.position - b.position);
            },
        },
        methods: {
            /**
             * Save current form state as new story
             */
            save() {
                this.$store.dispatch('stories/save', {
                    story: this.newStory
                }).then(() => {
                    this.newStory = {};
                });
            },

            /**
             * Save new position of dragged story
             * @param evt
             */
            saveOrder(evt) {
                const story = this.stories[evt.oldIndex];

                if (!story) {
                    return;
                }

                this.$store.dispatch('stories/patch', {
                    id: story.id,
                    field: 'position',
                    value: evt.newIndex + 1, // act_as_list is 1-indexed
                    fetch: true,
                });
            },
        }
    };
</script>

<style scoped>

</style>
