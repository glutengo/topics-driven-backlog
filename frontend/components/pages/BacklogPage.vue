<template>
    <section id="backlog-page">
        <h1>Backlog</h1>

        <table class="table table-striped">
            <thead>
                <tr>
                    <th>&nbsp;</th>
                    <th>Identifier</th>
                    <th>Story</th>
                    <th>Story&nbsp;points</th>
                    <th>Operations</th>
                </tr>
            </thead>
            <tbody v-sortable="{handle: '.js-drag-drop', onEnd: saveOrder}" >
                <tr 
                    is="StoryItem" 
                    v-for="story in stories" 
                    :key="story.id" 
                    :data="story" 
                    view="backlog"/>
            </tbody>
        </table>

        <StoryForm 
            v-if="showForm" 
            v-model="newStory" 
            @cancel="showForm = false" 
            @submit="save"/>
            
        <BButton 
            v-else 
            type="button" 
            variant="primary" 
            @click="showForm = true">Add story</BButton>
    </section>
</template>

<script>
import StoryItem from '@/components/elements/StoryItem';
import StoryForm from '@/components/forms/StoryForm';
import BButton from '@bootstrap/button/button';

export default {
    components: { StoryForm, StoryItem, BButton },
    data() {
        return {
            showForm: false,
            newStory: {},
        };
    },
    computed: {
        /**
         * All stories in backlog that are open and not already in sprints, sorted by position
         * @returns {array}
         */
        stories() {            
            return this.$store.getters['stories/all']
                .filter(s => !s.sprint_id && s.status === this.$store.state.stories.STATUS.OPEN)
                .sort((a, b) => a.position - b.position);
        },
    },
    methods: {
        /**
         * Save current form state as new story
         */
        save() {
            this.$store
                .dispatch('stories/save', {
                    story: this.newStory,
                })
                .then(() => {
                    this.newStory = {};
                });
        },

        /**
         * Save new position of dragged story
         * @param {Event} evt
         */
        saveOrder(evt) {
            const story = this.stories[evt.oldIndex];

            if (!story) {
                return;
            }

            this.$store.dispatch('stories/patch', {
                id: story.id,
                field: 'position',
                value: evt.newIndex + 1, // act_as_list is 1-indexed
                fetch: true,
            });
        },
    },
};
</script>