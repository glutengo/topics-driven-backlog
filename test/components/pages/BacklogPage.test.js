import { mount, createLocalVue } from 'vue-test-utils';
import { state } from '@/store/stories';
import Vuex from 'vuex';
import Vuelidate from 'vuelidate';
import BacklogPage from '@/components/pages/BacklogPage';
import StoryForm from '@/components/forms/StoryForm';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('BacklogPage.test.js', () => {
    let cmp, store, actions, getters;

    beforeEach(() => {
        actions = {
            save: jest.fn(),
            patch: jest.fn(),
        };

        getters = {
            all: () => [
                { id: 1, position: 2, status: 0 },
                { id: 2, position: 1, status: 0 },
                { id: 3, position: 3, status: 0, sprint_id: 2 },
                { id: 4, position: 4, status: 1 },
            ],
        };

        store = new Vuex.Store({
            modules: {
                stories: {
                    namespaced: true,
                    state: {
                        STATUS: state.STATUS,
                    },
                    actions,
                    getters,
                },
            },
        });

        cmp = mount(BacklogPage, {
            localVue,
            store,
        });
    });

    it('has sorted stories', () => {
        expect(cmp.vm.stories).toEqual([
            { id: 2, position: 1, status: 0 },
            { id: 1, position: 2, status: 0 },
        ]);
    });

    it('triggers story save when story form triggers submit', () => {
        cmp.vm.showForm = true;
        cmp.vm.newStory = { title: 'Test Story' };
        cmp.vm.save = jest.fn();
        cmp.update();

        cmp.find(StoryForm).trigger('submit');
        expect(cmp.vm.save).toBeCalled();
    });

    it('saves the story on story save', () => {
        cmp.vm.save();

        expect(actions.save).toBeCalled();
        expect(cmp.vm.newStory).toEqual({});
    });

    it('doesnt save the story order if story not found', () => {
        cmp.vm.saveOrder({ oldIndex: 2 });

        expect(actions.patch).not.toBeCalled();
    });

    it('saves the story order on sortable event', () => {
        cmp.vm.saveOrder({ oldIndex: 0, newIndex: 1 });

        expect(actions.patch).toBeCalled();
    });

    it('has the expected html structure', () => {
        expect(cmp.element).toMatchSnapshot();
    });

    it('has the expected html structure when form is displayed', () => {
        cmp.vm.showForm = true;
        cmp.update();

        expect(cmp.element).toMatchSnapshot();
    });
});
