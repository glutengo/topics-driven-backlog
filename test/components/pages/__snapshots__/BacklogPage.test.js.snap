// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BacklogPage.test.js has the expected html structure 1`] = `
<section
  id="backlog-page"
>
  <h1>
    Backlog
  </h1>
   
  <table
    class="table table-striped"
  >
    <thead>
      <tr>
        <th>
           
        </th>
         
        <th>
          Identifier
        </th>
         
        <th>
          Story
        </th>
         
        <th>
          Story points
        </th>
         
        <th>
          Operations
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="story-item"
      >
        <td>
          <span
            class="js-drag-drop"
          >
            <svg
              aria-label="Drag and drop to change order"
              class="fa-icon"
              height="16"
              role="img"
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19-19-45v-128h-384v384h128q26 0 45 19t19 45-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45 45-19h128v-384h-384v128q0 26-19 45t-45 19-45-19l-256-256q-19-19-19-45t19-45l256-256q19-19 45-19t45 19 19 45v128h384v-384h-128q-26 0-45-19t-19-45 19-45l256-256q19-19 45-19t45 19l256 256q19 19 19 45t-19 45-45 19h-128v384h384v-128q0-26 19-45t45-19 45 19l256 256q19 19 19 45z"
              />
               
    
              <!---->
            </svg>
          </span>
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-danger btn-sm"
            title="Delete"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M512 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM768 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM1024 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM480 384h448l-48-117q-7-9-17-11h-317q-10 2-17 11zM1408 416v64q0 14-9 23t-23 9h-96v948q0 83-47 143.5t-113 60.5h-832q-66 0-113-58.5t-47-141.5v-952h-96q-14 0-23-9t-9-23v-64q0-14 9-23t23-9h309l70-167q15-37 54-63t79-26h320q40 0 79 26t54 63l70 167h309q14 0 23 9t9 23z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
         
        <!---->
      </tr>
      <tr
        class="story-item"
      >
        <td>
          <span
            class="js-drag-drop"
          >
            <svg
              aria-label="Drag and drop to change order"
              class="fa-icon"
              height="16"
              role="img"
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19-19-45v-128h-384v384h128q26 0 45 19t19 45-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45 45-19h128v-384h-384v128q0 26-19 45t-45 19-45-19l-256-256q-19-19-19-45t19-45l256-256q19-19 45-19t45 19 19 45v128h384v-384h-128q-26 0-45-19t-19-45 19-45l256-256q19-19 45-19t45 19l256 256q19 19 19 45t-19 45-45 19h-128v384h384v-128q0-26 19-45t45-19 45 19l256 256q19 19 19 45z"
              />
               
    
              <!---->
            </svg>
          </span>
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-danger btn-sm"
            title="Delete"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M512 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM768 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM1024 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM480 384h448l-48-117q-7-9-17-11h-317q-10 2-17 11zM1408 416v64q0 14-9 23t-23 9h-96v948q0 83-47 143.5t-113 60.5h-832q-66 0-113-58.5t-47-141.5v-952h-96q-14 0-23-9t-9-23v-64q0-14 9-23t23-9h309l70-167q15-37 54-63t79-26h320q40 0 79 26t54 63l70 167h309q14 0 23 9t9 23z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
         
        <!---->
      </tr>
    </tbody>
  </table>
   
  <button
    class="btn btn-primary"
    type="button"
  >
    Add story
  </button>
</section>
`;

exports[`BacklogPage.test.js has the expected html structure when form is displayed 1`] = `
<section
  id="backlog-page"
>
  <h1>
    Backlog
  </h1>
   
  <table
    class="table table-striped"
  >
    <thead>
      <tr>
        <th>
           
        </th>
         
        <th>
          Identifier
        </th>
         
        <th>
          Story
        </th>
         
        <th>
          Story points
        </th>
         
        <th>
          Operations
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="story-item"
      >
        <td>
          <span
            class="js-drag-drop"
          >
            <svg
              aria-label="Drag and drop to change order"
              class="fa-icon"
              height="16"
              role="img"
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19-19-45v-128h-384v384h128q26 0 45 19t19 45-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45 45-19h128v-384h-384v128q0 26-19 45t-45 19-45-19l-256-256q-19-19-19-45t19-45l256-256q19-19 45-19t45 19 19 45v128h384v-384h-128q-26 0-45-19t-19-45 19-45l256-256q19-19 45-19t45 19l256 256q19 19 19 45t-19 45-45 19h-128v384h384v-128q0-26 19-45t45-19 45 19l256 256q19 19 19 45z"
              />
               
    
              <!---->
            </svg>
          </span>
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-danger btn-sm"
            title="Delete"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M512 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM768 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM1024 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM480 384h448l-48-117q-7-9-17-11h-317q-10 2-17 11zM1408 416v64q0 14-9 23t-23 9h-96v948q0 83-47 143.5t-113 60.5h-832q-66 0-113-58.5t-47-141.5v-952h-96q-14 0-23-9t-9-23v-64q0-14 9-23t23-9h309l70-167q15-37 54-63t79-26h320q40 0 79 26t54 63l70 167h309q14 0 23 9t9 23z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
         
        <!---->
      </tr>
      <tr
        class="story-item"
      >
        <td>
          <span
            class="js-drag-drop"
          >
            <svg
              aria-label="Drag and drop to change order"
              class="fa-icon"
              height="16"
              role="img"
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19-19-45v-128h-384v384h128q26 0 45 19t19 45-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45 45-19h128v-384h-384v128q0 26-19 45t-45 19-45-19l-256-256q-19-19-19-45t19-45l256-256q19-19 45-19t45 19 19 45v128h384v-384h-128q-26 0-45-19t-19-45 19-45l256-256q19-19 45-19t45 19l256 256q19 19 19 45t-19 45-45 19h-128v384h384v-128q0-26 19-45t45-19 45 19l256 256q19 19 19 45z"
              />
               
    
              <!---->
            </svg>
          </span>
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-danger btn-sm"
            title="Delete"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M512 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM768 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM1024 1376v-704q0-14-9-23t-23-9h-64q-14 0-23 9t-9 23v704q0 14 9 23t23 9h64q14 0 23-9t9-23zM480 384h448l-48-117q-7-9-17-11h-317q-10 2-17 11zM1408 416v64q0 14-9 23t-23 9h-96v948q0 83-47 143.5t-113 60.5h-832q-66 0-113-58.5t-47-141.5v-952h-96q-14 0-23-9t-9-23v-64q0-14 9-23t23-9h309l70-167q15-37 54-63t79-26h320q40 0 79 26t54 63l70 167h309q14 0 23 9t9 23z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
         
        <!---->
      </tr>
    </tbody>
  </table>
   
  <form
    class="story-form"
  >
    <div
      class="row"
    >
      <div
        class="col-md-10"
      >
        <div
          aria-labelledby="__BVID__99___BV_label_"
          class="b-form-group form-group"
          id="__BVID__99_"
          role="group"
        >
          <label
            class="col-form-label pt-0"
            for="story-title"
            id="__BVID__99___BV_label_"
          >
            Story *
          </label>
          <div
            class=""
          >
            <input
              aria-required="true"
              class="form-control"
              id="story-title"
              required="required"
              type="text"
            />
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
       
      <div
        class="col-md-2"
      >
        <div
          aria-labelledby="__BVID__101___BV_label_"
          class="b-form-group form-group"
          id="__BVID__101_"
          role="group"
        >
          <label
            class="col-form-label pt-0"
            for="story-points"
            id="__BVID__101___BV_label_"
          >
            Story points
          </label>
          <div
            class=""
          >
            <input
              class="form-control"
              id="story-points"
              min="0"
              type="number"
            />
            <!---->
            <!---->
            <!---->
          </div>
        </div>
      </div>
    </div>
     
    <div
      aria-labelledby="__BVID__103___BV_label_"
      class="b-form-group form-group"
      id="__BVID__103_"
      role="group"
    >
      <label
        class="col-form-label pt-0"
        for="story-description"
        id="__BVID__103___BV_label_"
      >
        Notes
      </label>
      <div
        class=""
      >
        <textarea
          class="form-control"
          id="story-description"
          rows="1"
          wrap="soft"
        />
        <!---->
        <!---->
        <!---->
      </div>
    </div>
     
    <button
      class="btn btn-primary disabled"
      disabled="disabled"
      type="submit"
    >
      Save
    </button>
     
    <button
      class="btn btn-grey"
      type="button"
    >
      Cancel
    </button>
  </form>
</section>
`;
